ext/gd/gd.c

```diff
@@ -1869,7 +1872,52 @@ static void _php_image_output(INTERNAL_FUNCTION_PARAMETERS, int image_type, char
 /* {{{ Output XBM image to browser or file */
 PHP_FUNCTION(imagexbm)
 {
-	_php_image_output_ctx(INTERNAL_FUNCTION_PARAM_PASSTHRU, PHP_GDIMG_TYPE_XBM, "XBM", gdImageXbmCtx);
+	zval *imgind;
+	char *file = NULL;
+	size_t file_len = 0;
+	zend_long foreground_color = -1;
+	gdImagePtr im;
+	int argc = ZEND_NUM_ARGS();
+	int i;
+	gdIOCtx *ctx = NULL;
+	php_stream *stream;
+
+	if (zend_parse_parameters(argc, "Op!|l", &imgind, gd_image_ce, &file, &file_len, &foreground_color) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	im = php_gd_libgdimageptr_from_zval_p(imgind);
+
+	if (argc > 1 && file != NULL) {
+		stream = php_stream_open_wrapper(file, "wb", REPORT_ERRORS|IGNORE_PATH|IGNORE_URL_WIN, NULL);
+		if (stream == NULL) {
+			RETURN_FALSE;
+		}
+
+		ctx = ecalloc(1, sizeof(gdIOCtx));
+		ctx->putC = _php_image_stream_putc;
+		ctx->putBuf = _php_image_stream_putbuf;
+		ctx->gd_free = _php_image_stream_ctxfreeandclose;
+		ctx->data = (void *)stream;
+	} else {
+		ctx = ecalloc(1, sizeof(gdIOCtx));
+		ctx->putC = _php_image_output_putc;
+		ctx->putBuf = _php_image_output_putbuf;
+		ctx->gd_free = _php_image_output_ctxfree;
```

Comment:
It seems that these 4 lines are repeated several times; what about extracting a function (something like `ctx = create_output_context()`)?

Reply:
Good idea! I could also extract a `create_stream_context()` function.